File: day-2.clj
(ns advent-of-code-2018.day-2)

(defn number-of-freq?
  [counts]
  (comp
   (partial < 0)
   count
   (partial filter (comp (partial = counts) second))))

(defn two-and-three-occurrences? [id]
  ((apply juxt
          (map number-of-freq?
               (list 2 3)))
   (frequencies id)))

(defn checksum [ids]
  (reduce
   *
   (map
    count
    ((juxt
      (partial filter first)
      (partial filter second))
     (map two-and-three-occurrences? ids)))))


;; === Test ========================================

(def ids
  (map seq ["aabbb" "aa" "x" "bbb"]))

(checksum ids)
;; => 4
;; as it should be, as there are two strings with pairs and two string with triples. Multiply these frequencies makes it four.

File: day_2.py
from collections import Counter
from functools import partial, reduce
from operator import mul

def occurring_frequencies(frequencies, items):
    counts = dict(Counter(items)).values()
    return set(f for f in frequencies if f in counts)

def checksum(items_list):
    frequencies = [2, 3]
    occurrences = [f for items in items_list
                     for f in occurring_frequencies(frequencies, items)]
    counts = Counter(occurrences).values()
    return reduce(mul, counts)

items_list = [list(items) for items in str.split("abc abbccc accc aa")]
print(checksum(items_list))
# => 4